feat: manage marketplace with miactl (#633)
mia-platform/documentation: docs/marketplace/add_to_marketplace/manage_marketplace_items.md

---
id: manage_marketplace_items
title:  Manage Marketplace Items
sidebar_label:  Manage Marketplace Items
---

There are three main methods to create, modify and delete Marketplace items:

* (recommended) Use [`miactl`](/cli/miactl/10_overview.md), the Mia-Platform command line interface tool.
* Open an issue on Mia-Platform [Github community page](https://github.com/mia-platform/community).
* (deprecated) Use the [CMS](/business_suite/guide_cms.md).

## Use `miactl`

First of all, you need to setup `miactl`, as explained in the [dedicated doc](/cli/miactl/20_setup.md).

With the `miactl marketplace` subcommands, you can perform several actions, described here below.

> ### Create an item 

Reviewer: I would put here below a brief introduction with a business use case (e.g. ACME Company decides to create an item on the Marketplace: why? --> link to another doc page. how? --> provide here below a practical example with Miactl)

:::info

> You need to have *Company Owner* or *Project Administrator* permission to perform this action

Reviewer: ```suggestion
You need to have *Company Owner* or *Project Administrator* role to perform this action
```

:::

First of all, you need to create a JSON file as explained in [this guide](/marketplace/add_to_marketplace/contributing_overview.md#how-to-configure-a-new-component). 

> Save the file in the current directory, for example as `myMarketplaceItem.json` file.

Reviewer: We may put an explicative screenshot of the "myMarketplaceItem.json" file

To create the item on the Marketplace, simply run this command:

```sh
miactl marketplace apply -f myMarketplaceItem.json
```

A message will confirm the operation, returning some information:
```
1 of 1 items have been successfully applied:

  ITEM ID                   NAME        STATUS   

>   65368hf0c91d871a87afbcbf  API Portal  Inserted  

Reviewer: We are now mentioning API Portal, but it was not mentioned above. Let's try to make a comprehensive and coherent use case narrative

```

If your object contains the `image` and/or the `supportedByImage` keys, such images will be uploaded along with the item.
The keys will be replaced with the `imageUrl` and the `supportedByImageUrl`, to obtain the updated version of the item use the `get` command: 
```sh
miactl marketplace get ITEM_ID > myMarketplaceItem.json

> ```

Reviewer: We can provide a picture of the item correctly created inside the Marketplace of the Console

More information on this point can be found in the [dedicated doc](/cli/miactl/30_commands.md#apply).

> ### Update an item

Reviewer: Why ACME Company should update it? Let's provide a brief example also in this case to keep the narration going

> More information on this point can be found in the [dedicated doc](/cli/miactl/30_commands.md#apply).
> ### Update an item
> 

Reviewer: ```suggestion
More information on this point can be found in the [dedicated doc](/cli/miactl/30_commands.md#apply).

### Update an item

```

:::info

> You need to have *Company Owner* or *Project Administrator* permission to perform this action

Reviewer: ```suggestion
You need to have *Company Owner* or *Project Administrator* role to perform this action
```

:::

To update an item already on the Marketplace you can download it and save it to a JSON file:

```sh
miactl marketplace get ITEM_ID > myMarketplaceItem.json
```

> where `ITEM_ID` is an alphanumerical id of the Marketplace element. If you don't know the item id, use the `miactl marketplace list` command to list all the Marketplace Items, you can easily locate the one of interest by looking for its name.

Reviewer: ```suggestion
where `ITEM_ID` is an alphanumerical id of the Marketplace element. If you don't know the item id, use the `miactl marketplace list` command to list all the Marketplace Items. You can easily locate the one of interest by looking for its name.
```

:::tip

It is suggested to always download the Marketplace Item just before updating it to be sure to work on the latest version.

:::

> Follow the steps in the [Modifying the Marketplace Item](#enabling-the-visibility-to-all-companies); once you are finished save the file and apply it to the Marketplace:

Reviewer: ```suggestion
Follow the steps in the [Modifying the Marketplace Item](#enabling-the-visibility-to-all-companies); once you have finished, save the file and apply it to the Marketplace:
```

```sh
miactl marketplace apply -f myMarketplaceItem.json
```

You will see the outcome of the operation in the command output:
```
1 of 1 items have been successfully applied:

  ITEM ID                   NAME        STATUS   

  65368hf0c91d871a87afbcbf  API Portal  Updated  
```

> ### Delete an item

Reviewer: Same as update

:::info

> You need either the *Company Owner* or *Project Administrator* permission to perform this action

Reviewer: ```suggestion
You need either the *Company Owner* or *Project Administrator* role at Company level to perform this action
```

---
id: manage_marketplace_items
title:  Manage Marketplace Items
sidebar_label:  Manage Marketplace Items
---

There are three main methods to create, modify and delete Marketplace items:

* (recommended) Use [`miactl`](/docs/cli/miactl/overview), the Mia-Platform command line interface tool.
* Open an issue on Mia-Platform [Github community page](https://github.com/mia-platform/community).
* (deprecated) Use the [CMS](/business_suite/guide_cms.md).

> ## Use `miactl`

Reviewer: Remember to point out what happens with Windows

Author: I would not do it here, I would do it in miactl docs, along with the info of the other OSes. I'll open a PR shortly on miactl repo

Author: Here it is mia-platform/miactl#105

---
id: manage_marketplace_items
title:  Manage Marketplace Items
sidebar_label:  Manage Marketplace Items
---

There are three main methods to create, modify and delete Marketplace items:

* (recommended) Use [`miactl`](/cli/miactl/10_overview.md), the Mia-Platform command line interface tool.
* Open an issue on Mia-Platform [Github community page](https://github.com/mia-platform/community).
* (deprecated) Use the [CMS](/business_suite/guide_cms.md).

## How to use `miactl` to manage the Marketplace

First of all, you need to setup `miactl`, as explained in the [dedicated doc](/cli/miactl/20_setup.md).

With the `miactl marketplace` subcommands, you can perform several actions, described here below.

### Create an item 

:::info

You need to have *Company Owner* or *Project Administrator* role at Company level to perform this action

:::

Imagine you are a software developer, working for the Company "Acme Corporation".

You have developed a new service (for example a NodeJS service) 

> called "My Awesome Service",and you want it to be available in the Marketplace of your Company.

Reviewer: ```suggestion
called "My Awesome Service", and you want it to be available in the Marketplace of your Company.
```

First of all, you need to create a JSON file as explained in [this guide](/marketplace/add_to_marketplace/contributing_overview.md#how-to-configure-a-new-component). 

> Save the file in the current directory, for example as `myMarketplaceItem.json` file.

Reviewer: i would substitute the json file name with the new name

> Save the file in the current directory, for example as `myAwesomeService.json` file.

Reviewer: I would talk about the current directory below, when we introduce the command

> The file contents will probably look like this:

Reviewer: ```suggestion
The file content will probably look like this:
```

```json
{
  "description": "My Awesome Service allows your project to do amazing stuff!",
  "documentation": {
    "type": "externalLink",
    "url": "https://docs.example.org/AwesomeService"
  },
  "image": {
    "localPath": "./awesomeService.png"
  },
  "name": "My Awesome Service",
  "publishOnMiaDocumentation": true,
  "repositoryUrl": "https://git.example.org/awesome-service",
  "resources": {
    "services": {
      "api-portal": {

>         "componentId": "api-portal",

Reviewer: I would remove the reference to API Portal

        "containerPorts": [
          {
            "from": 80,
            "name": "http",
            "protocol": "TCP",
            "to": 8080
          }
        ],
        "defaultEnvironmentVariables": [
          {
            "name": "HTTP_PORT",
            "value": "8080",
            "valueType": "plain"
          }
        ],
        "defaultLogParser": "mia-nginx",
        "defaultProbes": {
          "liveness": {
            "path": "/index.html"
          },
          "readiness": {
            "path": "/index.html"
          }
        },
        "defaultResources": {
          "memoryLimits": {
            "max": "25Mi",
            "min": "5Mi"
          }
        },
        "description": "My Awesome Service allows your project to do amazing stuff!",
        "dockerImage": "docker.example.org/awesome-service:1.0",
        "name": "awesome-service",
        "repositoryUrl": "https://git.example.org/awesome-service",
        "type": "plugin"
      }
    }
  },

>   "supportedBy": "Mia-Platform",

Reviewer: Mia-Platform or Acme Company?

  "supportedByImage": {
    "localPath": "./acmeCorporation.png"
  },

>   "tenantId": "team-rocket-test",

Reviewer: shouldn't be something like "acme-company" ?

  "type": "plugin"
}
```

> Notice that the `image` and `supportedByImage` objects are populated with local paths to images: make sure the images exists and that their path is correct.

Reviewer: ```suggestion
Notice that the `image` and `supportedByImage` objects are populated with local paths to images: make sure the images exist and that their path is correct.
```

To create the item on the Marketplace, simply run this command:

```sh
miactl marketplace apply -f myAwesomeService.json
```

This command will create the Marketplace item and upload the images along with it.

> A message will confirm the operation, returning some information:

Reviewer: ```suggestion
A message will confirm the operation, returning some information as shown here below:
```

1 of 1 items have been successfully applied:

>   ITEM ID                   NAME                 STATUS   

Reviewer 2: ```suggestion
  ID                  NAME                 STATUS   
```

  ITEM ID                   NAME                 STATUS   

  65368hf0c91d871a87afbcbf  My Awesome Service   Inserted  
```

> After the upload, the image keys will be replaced with the `imageUrl` and the `supportedByImageUrl`; to obtain the updated version of the item use the `get` command: 

Reviewer: ```suggestion
After the upload, the image keys will be replaced with the `imageUrl` and the `supportedByImageUrl`; to obtain the updated version of the item, use the `get` command: 
```

```sh
miactl marketplace get 65368hf0c91d871a87afbcbf > myAwesomeService.json
```

:::tip

The local file fields won't be updated after the item creation.
Make sure to always download a new copy afterwards to keep your local copy up to date.

:::

> From now on, the service "My Awesome Service" will be visible in the Marketplace section of the Console.

Reviewer: ```suggestion
From now on, the service "My Awesome Service" will be visible as a clickable card inside the Marketplace section of the Console.
```

![awesome service](img/awesome_service.png)

:::tip

Further information about the `apply` command can be found in the [dedicated doc](/cli/miactl/30_commands.md#apply).

:::

### Update an item

:::info

You need to have *Company Owner* or *Project Administrator* role at Company level to perform this action

:::

Imagine now that you noticed that the description of "My Awesome Service" is not correct and you want to change it.

First of all, download and save the latest version of the item configuration:

```sh
miactl marketplace get ITEM_ID > myAwesomeService.json
```

> where `ITEM_ID` is an alphanumerical id of the Marketplace item. If you don't know the item id, use the `miactl marketplace list` command to list all the Marketplace Items. You can easily locate the one of interest by looking for its name.

Reviewer: ```suggestion
where `ITEM_ID` is an alphanumerical id of the Marketplace item.  
If you don't know the item id, use the `miactl marketplace list` command to list all the Marketplace Items. You can easily locate the one of interest by looking for its name.
```

> where `ITEM_ID` is an alphanumerical id of the Marketplace element. If you don't know the item id, use the `miactl marketplace list` command to list all the Marketplace Items. You can easily locate the one of interest by looking for its name.

Reviewer: ```suggestion
where `ITEM_ID` is an alphanumerical id of the Marketplace item. If you don't know the item id, use the `miactl marketplace list` command to list all the Marketplace Items. You can easily locate the one of interest by looking for its name.
```

:::tip

> It is suggested to always download the Marketplace Item just before updating it to be sure to work on the latest version.

Reviewer: ```suggestion
It is suggested to always download the Marketplace item just before updating it to make sure it works on the latest version.
```

:::

> Edit your file following the steps in the [Modifying the Marketplace Item](#enabling-the-visibility-to-all-companies); 

Reviewer: ```suggestion
Edit your file following the steps described in the [Modifying the Marketplace Item](#enabling-the-visibility-to-all-companies); 
```

once you are happy with the changes, save the file and apply it to the Marketplace:

```sh
miactl marketplace apply -f myMarketplaceItem.json
```

You will see the outcome of the operation in the command output:
```
1 of 1 items have been successfully applied:

  ITEM ID                   NAME                STATUS   

  65368hf0c91d871a87afbcbf  My Awesome Service  Updated  
```

The changes are now reflected to the Console.

### Delete an item

:::info

You need either the *Company Owner* or *Project Administrator* role at Company level to perform this action

:::

> Imagine notice that "My Awesome Service" is not useful anymore, and you want to delete it.

Reviewer: ```suggestion
Imagine you notice that the service "My Awesome Service" is no longer useful for your Company and so you want to delete it.
```

You can delete an item from the Marketplace by means of the `delete` command:

```sh
miactl marketplace apply -f myAwesomeService.json
```

The item is then deleted from the Marketplace. 

The deletion is permanent, but the file on your machine will not be deleted.

> If you want, you can recreate the item on the Marketplace again by applying the file.
> 
> 
> 
> ## Open an issue on Mia-Platform Github community page
> 

Reviewer: ```suggestion
If you want, you can recreate the item on the Marketplace again by applying the file.

## Open an issue on Mia-Platform Github community page

```
